test/test_inverse.py

```diff
@@ -177,10 +181,170 @@ def test_KFAC_inverse_damped_matmat(
+        if ggT is not None:
+            ggT.add_(torch.eye(ggT.shape[0], device=ggT.device), alpha=damping_ggT)
+
+    # manual heuristically damped inverse
+    inv_KFAC_naive = torch.inverse(torch.as_tensor(KFAC @ eye(KFAC.shape[0])))
+
+    # remove heuristic damping
+    for mod_name in KFAC._mapping.keys():
+        aaT = KFAC._input_covariances.get(mod_name)
+        ggT = KFAC._gradient_covariances.get(mod_name)
+        damping_aaT, damping_ggT = heuristic_damping.get(mod_name, (delta, delta))
+        if aaT is not None:
+            aaT.sub_(torch.eye(aaT.shape[0], device=aaT.device), alpha=damping_aaT)
+        if ggT is not None:
+            ggT.sub_(torch.eye(ggT.shape[0], device=ggT.device), alpha=damping_ggT)
+
+    # check that passing a tuple for heuristic damping will fail
+    with raises(ValueError):
+        inv_KFAC = KFACInverseLinearOperator(
+            KFAC, damping=(delta, delta), use_heuristic_damping=True
+        )
+
+    # use heuristic damping with KFACInverseLinearOperator
+    inv_KFAC = KFACInverseLinearOperator(
+        KFAC,
+        damping=delta,
+        cache=cache,
+        use_heuristic_damping=True,
+        min_damping=KFAC_MIN_DAMPING,
+    )
+
+    num_vectors = 2
+    X = random.rand(KFAC.shape[1], num_vectors)
+    # test for equivalence
+    report_nonclose(inv_KFAC @ X, inv_KFAC_naive @ X, rtol=5e-2)
+
+    assert inv_KFAC._cache == cache
+    if cache:
+        # test that the cache is not empty
+        assert len(inv_KFAC._inverse_input_covariances) > 0
+        assert len(inv_KFAC._inverse_gradient_covariances) > 0
+    else:
+        # test that the cache is empty
+        assert len(inv_KFAC._inverse_input_covariances) == 0
+        assert len(inv_KFAC._inverse_gradient_covariances) == 0
+
+
+@mark.parametrize("cache", [True, False], ids=["cached", "uncached"])
+@mark.parametrize(
+    "exclude", [None, "weight", "bias"], ids=["all", "no_weights", "no_biases"]
+)
+@mark.parametrize(
+    "separate_weight_and_bias", [True, False], ids=["separate_bias", "joint_bias"]
+)
+def test_KFAC_inverse_exactly_damped_matmat(
+    case, cache: bool, exclude: str, separate_weight_and_bias: bool, delta: float = 1e-2
+):
+    """Test matrix-matrix multiplication by an inverse (exactly) damped KFAC approximation."""
+    model_func, loss_func, params, data = case
+
+    if exclude is not None:
+        names = {p.data_ptr(): name for name, p in model_func.named_parameters()}
+        params = [p for p in params if exclude not in names[p.data_ptr()]]
+
+    loss_average = "batch" if loss_func.reduction == "mean" else None
+    KFAC = KFACLinearOperator(
+        model_func,
+        loss_func,
+        params,
+        data,
+        loss_average=loss_average,
+        separate_weight_and_bias=separate_weight_and_bias,
+    )
+
+    # manual exactly damped inverse
+    inv_KFAC_naive = torch.inverse(
+        KFAC.torch_matmat(torch.eye(KFAC.shape[0])) + delta * torch.eye(KFAC.shape[0])
+    )
```

Comment:
Is there a reason to use `torch_matmat` over the public `matmat` with a `numpy` array?

Reply:
I am confused because in L346 `X` is a `numpy` array, but `inv_KFAC_naive` seems to be a `torch.Tensor` that is multiplied onto it. I expect this should crash if we execute the tests on a machine with CUDA device.

Reply:
True, I added `.cpu().numpy()` in some places. And there is no reason for using `torch_matmat` besides that I computed the inverse in `PyTorch`.